_posts/2016-03-12-make-your-data-obey.markdown

```diff
@@ -0,0 +1,187 @@
+.then(db.create).then(res => {
+    console.log('User Created!', res)
+})
+.catch(err => {
+    console.log('Failed!', err)
+})
+{% endhighlight %}
+
+Overall pretty simple stuff. We validate the model, then create the user (or catch the error). But what about that `id`? We probably want to have that generated automatically. Yes, most databases auto-generate unique id's, but play along; this is a perfect example.
+
+## Creators
+
+Obey supplies `creators` for generating data using methods that can be predefined. For the `id` we'll use the [node-uuid](https://www.npmjs.com/package/node-uuid) module. Let's revisit the model:
+
+{% highlight javascript %}
+import obey from 'obey'
+// Import the node-uuid lib
+import uuid from 'node-uuid'
+
+// Add a creator that returns a v4 UUID if one isn't supplied
+obey.creator('uuidCreator', () => uuid.v4())
+
+const user = obey.model({
+    // The id now specifies that the creator 'uuid' should be used if no value is present
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    password: { type: 'string', require: true }
+})
+{% endhighlight %}
+
+So, now we just pass in the `email` and `password` and our `id` is auto-generated.
+
+But that `password` is still not cool; we don't want to save that in plain-text...
+
+## Modifiers
+
+Obey wasn't just built for validation, but also to support coercion and modification. Modifiers allow for doing this in a clean, simple method. Let's revisit the model again:
+
+{% highlight javascript %}
+import obey from 'obey'
+import uuid from 'node-uuid'
+// Import the crypto module
+import crypto from 'crypto'
+
+obey.creator('uuid', () => uuid.v4())
+
+// Add a modifier that returns md5 encrypted value
+obey.modifier('encrypt', val => crypto.createHash('md5').update(val).digest('hex'))
+
+const user = obey.model({
+    id: { type: 'uuid', creator: 'uuidCreator' required: true },
+    email: { type: 'email', required: true },
+    // Password now specifies the modifier 'encrypt' should be used on the value
+    password: { type: 'string', modifier: 'encrypt', require: true }
+})
+{% endhighlight %}
+
+Alright, so now not only does our validation create a UUID, it also encrypts that password for us. We've changed the model to accomplish this, but we validate exactly the same, Obey handles this all automatically based on the modelling we defined in one location.
+
+BUT, what about that `email`? Probably want to make sure it's unique...
+
+## Types
+
+Having that `email` contain one that's already in the datasource would be bad news bears right? So here's the point where a lot of developers say "my database will reject it!". Good point, not so good execution though.
+
+You want your datasource to go into (a potentially locking) `insert` state to do your validation for you? Really? Why not do that all inside the validation? It is validating data after all. Luckily `types` are customizable, so let's make our own:
```

Comment:
Opening yourself up for huge argument here-- if you have an index on the email, there's no perceptible lock under practically any DB. I'd change this up to focus on A distributed DB that can't provide uniqueness integrity, outside of SQLs there are a lot of these-- and even shared SQLs make this difficult.